### PR TITLE
[#136] Bump default tarball_max_size from 8 MiB to 16 MiB

### PR DESCRIPTION
**Pull request by onno-vos-dev**
I opted for the 2nd option as per described in the issue as it's the quickest and easiest to resolve my issue and allow me to publish docs again. I'll open a separate issue on `rebar3_hex` to discuss the option of moving to `create_docs/2`.

### CODE REVIEW

File: src/hex_core.erl [hunk 1]
```diff
@@ -102,6 +102,6 @@ default_config() ->
         repo_organization => undefined,
         repo_verify => true,
         repo_verify_origin => true,
-        tarball_max_size => 8 * 1024 * 1024,
+        tarball_max_size => 64 * 1024 * 1024,
```
Comment (ericmj) [inline src/hex_core.erl:105]:
```suggestion
        tarball_max_size => 16 * 1024 * 1024,
```

**Comment by wojtekmach** [comment]
Thank you for the PR. This change will up the limit of _package_ tarballs too which I'm not opposed to in general but I'd wait for a concrete use case. If all you need is bigger _docs_ tarball size, I'd rather focus on that.

Changing rebar3_hex is the way to go. In the meantime you can manually publish the docs using this elixir script:

```elixir
Mix.install([:req])

path = "docs.tar.gz"
{_, 0} = System.cmd("rebar3", ~w(ex_doc), into: IO.stream())
{_, 0} = System.cmd("tar", ~w(-czf #{path} --cd doc .), into: IO.stream())
tarball = File.read!(path)
File.rm_rf!(path)

# $ mix hex.user key generate --permission api:write
token = "..."
package = "hex_core"
version = "0.9.0"

Req.post!("https://hex.pm/api/packages/#{package}/releases/#{version}/docs",
  body: tarball,
  headers: [authorization: token]
)
|> dbg()
```

It's not a fantastic solution but hopefully it unblocks you in the meantime?

**Comment by starbelly** [comment]
@wojtekmach  What's the limit for doc tarballs on the server side? I believe saw 10MB, can you confirm that is correct?

**Comment by onno-vos-dev** [comment]
@wojtekmach I can try out the Elixir script you posted later but I'd still argue for upping the limit as I'm fairly sure `aws_erlang` will breach the 8MiB limit fairly soon. It currently sits at around 6.8 MiB and considering that each time AWS opens a new API or expands on existing API's/docs, `aws_erlang` simply automatically generates the code for it, `aws_erlang` is essentially only getting bigger and rarely shrinks (AWS rarely deprecates services). So it's mostly a matter of time before we reach the limit. 
 
 Is there any chance you can re-consider? :-) Otherwise I'll push for the change in `rebar3_hex` and try and spend some time on the weekend to make a PR there :-)

**Review by wojtekmach — APPROVED** [review]
> I'd still argue for upping the limit as I'm fairly sure aws_erlang will breach the 8MiB limit fairly soon-ish. It currently sits at around 6.8 MiB 

oh, ok, thanks for letting us know!

**Comment by wojtekmach** [comment]
@starbelly yes at the moment hexpm only allows request bodies with ~10mb.

**Comment by onno-vos-dev** [comment]
@wojtekmach What would be required to have this merged? :-) Is there anything I can do? :-)

**Comment by ericmj** [comment]
hex.pm has been updated to allow docs tarballs up to 16MiB size.

File: src/hex_erl_tar.erl [hunk 2]
```diff
@@ -84,6 +84,8 @@ format_error({invalid_gnu_0_1_sparsemap, Format}) ->
     lists:flatten(io_lib:format("Invalid GNU sparse map (version ~s)", [Format]));
 format_error(unsafe_path) ->
     "The path points above the current working directory";
+format_error(too_big) ->
+    "The tar exceeds tarball_max_size or tarball_max_uncompressed_size";
```
[the remark this answers is not in the document]
Comment (starbelly) [reply src/hex_erl_tar.erl:88]: The error should be handled inside rebar3_hex, but I see a gotcha with it, I'll do a fixup there, the error there would look like : 

```shell
===> Error building package : package exceeds max compressed size 8.388608 MB
```

[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply src/hex_erl_tar.erl:88]: I will revert the error from this PR :+1:

**Comment by starbelly** [comment]
@ericmj To unblock @onno-vos-dev should I just precede with cutting a version of rebar3_hex passing in the new size for docs tarball?

File: src/hex_core.erl [hunk 3]
```diff
@@ -102,6 +102,6 @@ default_config() ->
         repo_organization => undefined,
         repo_verify => true,
         repo_verify_origin => true,
-        tarball_max_size => 8 * 1024 * 1024,
+        tarball_max_size => 16 * 1024 * 1024,
         tarball_max_uncompressed_size => 64 * 1024 * 1024
```
Comment (ericmj) [inline src/hex_core.erl:106]:
```suggestion
        tarball_max_uncompressed_size => 128 * 1024 * 1024
```

**Pull request closed without merging**